Proposed API change

jfrolich opened the pull request:
This is a WIP, but I found that the functor API created quite a bit of boilerplate code. When I looked at the source of [reason-urql](https://github.com/FormidableLabs/reason-urql), it found a way to implement a more concise API, so I messed around with the source of reason-react-hooks if we can have something like that that with apollo and this is the result. I would like to get your opinion and possible improvements.

In short we can go from this:
```reason
module ActivityPhotosQueryConfig = [%graphql
  {|
  query ActivityPhotos($feedId: ID, $scaleFactor: Float!, $height: Float!) {
    feedImageList(feedId: $feedId) {
      nodes {
        id
        width
        height
        file(scaleFactor: $scaleFactor, height: $height) {
          url
        }
      }
    }
  }
|}
];
module ActivityPhotosQuery =
  ReasonApolloHooks.Query.Make(ActivityPhotosQueryConfig);

[@react.component]
let make = (~feedId=?, ~small=false, ~edit=false) => {
  let height = small ? 100. : 200.;
  let (query, _full) =
    ActivityPhotosQuery.use(
      ~variables=
        ActivityPhotosQueryConfig.make(
          ~feedId?,
          ~scaleFactor=PixelRatio.get(),
          ~height,
          (),
        )##variables,
      (),
    );
}
```

to this:

```reason
module ActivityPhotosQuery = [%graphql
  {|
  query ActivityPhotos($feedId: ID, $scaleFactor: Float!, $height: Float!) {
    feedImageList(feedId: $feedId) {
      nodes {
        id
        width
        height
        file(scaleFactor: $scaleFactor, height: $height) {
          url
        }
      }
    }
  }
|}
];


[@react.component]
let make = (~feedId=?, ~small=false, ~edit=false) => {
  let height = small ? 100. : 200.;
  let (query, _full) =
    ReasonApolloHooks.useQuery(
      ~query=
        ActivityPhotosQuery.make(
          ~feedId?,
          ~scaleFactor=PixelRatio.get(),
          ~height,
          (),
        ),
      (),
    );
}
```

Similar for mutations, however we need to pass in an initial mutation if we want to provide variables later:

```reason
let (create, _simple, _full) =
  ReasonApolloHooks.useMutation(
    ~incompleteMutation=CreateFeedImageMutation.query,
  (),
);
```

We can run the create function like this:

```reason
create(
  ~mutation=CreateFeedImageMutation.make(~feedImage, ()),
  ~refetchQueries=_ => [|refetchQuery(~feedId?, ())|],
  ~awaitRefetchQueries=true,
  (),
)
```

or we can add the mutation from the get go (and then we don't have to supply it later):

```reason
let (create, _simple, _full) =
  ReasonApolloHooks.useMutation(
    ~mutation=CreateFeedImageMutation.make(~feedImage, ()),
  (),
);
```

To be honest the mutation API is suboptimal. I think there is a way that we do not need to provide the `incompleteMutation` but this involves changing the javascript `useMutation` hook.
OR we need to modify `graphql_ppx`, to generate a different module.

fakenickels commented:
Thanks for your proposal! 
I think we can leave this opened for a while to see what others will say about it

jfrolich commented:
Sounds great. I am pretty new to Reason/OCaml, so there are probably many things that can be better.

fakenickels commented:
No opinions of otherss so far but in overall I liked the simplification.
Can you update the examples? After that I think we are good to go

jfrolich commented:
I updated the PR, a few changes:
- Changed the main module to ApolloHooks (debatable, but more concise it's possible now that the package is not namespaced anymore)
- Updated `graphql_ppx` in the example to `graphql_ppx_re` (which is a maintained version of `graphql_ppx` in reason -- same API).
- Removed the namespacing so we can do `ApolloHooks.useQuery` directly instead of `ApolloHooks.Query.useQuery` or something. Also have all the hooks available when opening `ApolloHooks`.

fakenickels commented:
Thanks a lot @jfrolich ! These next weeks we'll be release busy for me but I'll try to review and merge ASAP

fakenickels commented on examples/persons/package.json:
```diff
@@ -19,17 +19,17 @@
+    "bs-platform": "^5.2.0",
     "css-loader": "^3.2.0",
-    "graphql_ppx": "^0.2.8",
+    "@baransu/graphql_ppx_re": "^0.2.0-beta.2",
```
very good!

jfrolich replied:
Yeah so cool to see activity again in this package. It has so much potential.

jfrolich commented:
> Thanks a lot @jfrolich ! These next upcoming weeks we'll be really busy for me but I'll try to review and merge ASAP

Cool. No worries. When this is merged I also have a pull-request that allows to do `fetchMore` in pure reason, instead of the raw bucklescript escape hatch!

sgrove commented:
This looks fantastic!

We've added code-gen support for the current API, you can see the current usage for both queries and mutations here: https://serve.onegraph.com/short/B8DR68 (open code exporter and choose `Reason` as the language)

I'd love to get this merged in and make the code that much more approachable for workshops (and for me too of course!).

fakenickels commented:
Amazing!! Will get this merged before weekend 🙏

jfrolich commented:
> This looks fantastic!
> 
> We've added code-gen support for the current API, you can see the current usage for both queries and mutations here: https://serve.onegraph.com/short/B8DR68 (open code exporter and choose `Reason` as the language)
> 
> I'd love to get this merged in and make the code that much more approachable for workshops (and for me too of course!).

Wow that codegen is awesome!

jfrolich commented:
> Amazing!! Will get this merged before weekend 🙏

I merged in the upstream changes and upgraded `graphql_ppx_re` in the example.

fakenickels commented:
Merged will release soon!

fakenickels commented:
Thanks lot @jfrolich for this!! Awesome work. And thanks everyone for giving feedback

MargaretKrutikova commented:
I am a bit late to the party, but what happened to `errorPolicy` in `useQuery`? I think it disappeared.

fakenickels commented:
Looks like we have some regressions here indeed, along with #57

MargaretKrutikova commented on src/ApolloHooksTypes.re:
```diff
@@ -43,18 +44,4 @@ let fetchPolicyToJs = fetchPolicy =>
-/**
- * apollo-client/src/core/watchQueryOptions.ts
- */
-type errorPolicy =
```
Are those not used anymore?

jfrolich replied:
Ah that is probably an upstream change that I forgot to merge in (diffs were not working anymore so I had to manually look up what has been added to master in the meantime)

fakenickels commented:
We haven't released this just yet so there is still time to fix

MargaretKrutikova commented:
I will try to run the example project locally, and see what I can fix 😄

MargaretKrutikova commented on src/ApolloHooksMutation.re:
```diff
@@ -0,0 +1,210 @@
+          | (None, None) => raise(Error("Need to pass a mutation"))
+          };
+
+        React.Ref.setCurrent(parse, Some(mutation##parse));
```
Oh, this runs inside of `useMemo`? Callback passed into `useMemo` must be pure because it will run on every render, see [react docs](`https://reactjs.org/docs/hooks-reference.html#usememo`). And later it is called `jsMutate` which will send the actual query, is it correct?
 I am not sure it is a good idea to implement it this way, I think we will have to rethink this approach a bit, what do you think, @fakenickels ?

MargaretKrutikova replied:
Sorry, my bad, spent some time alazying the code and realised that it is memoizing the function and not actually calling it 🤦‍♀ 😄

jfrolich replied:
I merged this change from master, but good suggestion. useCallback is just a shortcut for memoizing functions with useMemo if I remember correctly.

MargaretKrutikova commented on src/ApolloHooksMutation.re:
```diff
@@ -0,0 +1,210 @@
+
+let useMutation =
+    (
+      ~incompleteMutation=?,
```
I am not sure `incompleteMutation` is the right name here, it might get really verbose and tedious to write on many mutations, can we come up with something more succinct for this case? 😄

jfrolich replied:
Usually you would use the mutation argument, unless you need to set the arguments dynamically. It would be better to have the mutation as an optional argument, but that needs some changes on the javascript side (I think!). Anyway I agree it's not ideal (and hopefully temporarily). Let me know what better name you are thinking of!

MargaretKrutikova commented on src/ApolloHooksMutation.re:
```diff
@@ -0,0 +1,210 @@
+        | (_, Some(incompleteMutation)) => incompleteMutation
+        | (None, None) =>
+          raise(
+            Error("You need to provide a mutation or an incomplete mutation"),
```
I think we need to have a better exception type, something more descriptive than `Error`, does it makes sense, @fakenickels ?

MargaretKrutikova replied:
Maybe we can avoid throwing errors here by making sure the user can't make a mistake on the type level? I think it will be better to not throw errors if the type system can guarantee that the values are correct 😄 I was thinking about using some kind of a variant:
```
type mutationConfig('a) = Query(string) | Config('a);
```
and then have it in `useMutation`:
```
let useMutation =
    (
      ~mutation: mutationConfig('a)=?,
      ~client=?,
     ....
      (),
    )
```

jfrolich replied:
Ideally we can remove the need to pass in incompleteMutation in the future which will solve this issue.

fakenickels commented:
I think will be better to revert this and open the PR again so we can make we have less breaking changes

fakenickels commented:
@jfrolich would you mind reopening this PR?

jfrolich commented:
Yep. Will re-open. So if I read correctly I missed the following upstream changes in master in this PR:
- [ ] skip parameter
- [ ] errorPolicy

The other comments of @MargaretKrutikova are about existing code not related to this PR. This can be tackled in a separate PR. (Except the naming of `incompleteMutation`, happy to change this name if anyone has a name that is better suited or a way so we don't need to pass in the incomplete mutation.

fakenickels commented:
@all-contributors add @jfrolich code, review and ideas
